Nimble/Utils/NotificationCollector.swift

```diff
@@ -0,0 +1,50 @@
+//
+//  NotificationCollector.swift
+//  Nimble
+//
+//  Created by Brian Gerstle on 12/31/15.
+//  Copyright © 2015 Jeff Hui. All rights reserved.
+//
+
+import Foundation
+
+class NotificationCollector {
+
+    private var observedNotifications: [NSNotification]
+    private let notificationCenter: NSNotificationCenter
+    private var token: AnyObject?
+
+    required init(notificationCenter: NSNotificationCenter) {
+        self.notificationCenter = notificationCenter
+        self.observedNotifications = []
+    }
+
+    func startObserving() {
+        self.token = self.notificationCenter.addObserverForName(nil, object: nil, queue: nil) { [weak self] n -> Void in
+            self?.observedNotifications.append(n)
+        }
+    }
+
+    deinit {
+        if let token = self.token {
+            self.notificationCenter.removeObserver(token)
+        }
+    }
+}
+
+public typealias CollectedNotificationsFunc = () throws -> [NSNotification]?
+
+public func postedNotifications(notificationCenter: NSNotificationCenter = NSNotificationCenter.defaultCenter(),
+                                block: () throws -> Void) -> CollectedNotificationsFunc {
+    let collector = NotificationCollector(notificationCenter: notificationCenter)
+    collector.startObserving()
+    var once: Bool = false
+    return {
+        assert(NSThread.isMainThread(), "Only expecting closure to be evaluated on main thread.")
+        if !once {
+            once = true
+            try block()
+        }
+        return collector.observedNotifications
+    }
+}
```

Comment:
I think the `postedNotifications` func is causing problems that forces `expect(expression: ...)` form you're talking about. `postedNotification` should return the observed notifications:

```swift
public func postedNotifications(notificationCenter: NSNotificationCenter = NSNotificationCenter.defaultCenter(),
                                block: () throws -> Void) throws -> [NSNotification]
```

[Nimble's expect(...)](https://github.com/Quick/Nimble/blob/master/Nimble/DSL.swift#L2-L17) argument is an [@autoclosure](https://developer.apple.com/library/ios/documentation/Swift/Conceptual/Swift_Programming_Language/Attributes.html). This means the expression in that argument is an implicit closure:

```swift
// this line
expect(postedNotification { ... }).to(equal(...))
// is equivalent to this line
expect { postedNotification { ... } }.to(equal(...))
```

This allows Nimble to catch errors despite looking like a normal argument:

```swift
expect(try somethingThatThrows()).to(throwError(...))
expect(raisesNSException()).to(raiseException(...))
```